modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java

```diff
@@ -0,0 +1,403 @@
+        restClient = new ActionsRestClient(serverURL, tenantInfo);
+        // TODO: Review if ActionsRestClient should be instantiated, or if the superclass initialization is sufficient
+
+        List<String> customScopes = Arrays.asList(CUSTOM_SCOPE_1, CUSTOM_SCOPE_2, CUSTOM_SCOPE_3);
+
+        ApplicationResponseModel application = addApplicationWithGrantType(PASSWORD_GRANT_TYPE);
+        if (!CarbonUtils.isLegacyAuthzRuntimeEnabled()) {
+            authorizeSystemAPIs(application.getId(), new ArrayList<>(Arrays.asList(SCIM2_USERS_API, ACTIONS_API,
+                    APPLICATION_MANAGEMENT_API, API_RESOURCE_MANAGEMENT_API)));
+        }
+        String domainAPIId = createDomainAPIs(EXTERNAL_SERVICE_NAME, EXTERNAL_SERVICE_URI, customScopes);
+        authorizeDomainAPIs(application.getId(), domainAPIId, customScopes);
+
+        addUserWithRole(application.getId(), customScopes);
+
+        MockServer.createMockServer(MOCK_SERVER_ENDPOINT);
+        createPreIssueAccessTokenAction();
+
+        accessToken = retrieveAccessToken(application.getId(), customScopes);
+        jwtClaims = extractJwtClaims(accessToken);
+    }
+
+    @Test(groups = "wso2.is", description = "Verify the presence of the added scope in the access token")
+    public void testTokenScopeOperations() throws Exception {
+
+        String[] scopes = jwtClaims.getStringClaim("scope").split("\\s+");
+
+        Assert.assertTrue(ArrayUtils.contains(scopes, NEW_SCOPE_1));
+        Assert.assertTrue(ArrayUtils.contains(scopes, NEW_SCOPE_4));
+        Assert.assertFalse(ArrayUtils.contains(scopes, CUSTOM_SCOPE_3));
+        Assert.assertFalse(ArrayUtils.contains(scopes, CUSTOM_SCOPE_2));
+    }
+
+    @Test(groups = "wso2.is", description = "Verify the presence of the added scope in the access token")
+    public void testTokenAUDClaimOperations() throws Exception {
+
+        String[] audValueArray = jwtClaims.getStringArrayClaim("aud");
+
+        Assert.assertTrue(ArrayUtils.contains(audValueArray, "zzz1.com"));
+        Assert.assertTrue(ArrayUtils.contains(audValueArray, "zzzR.com"));
+        Assert.assertFalse(ArrayUtils.contains(audValueArray, clientId));
+    }
+
+    @DataProvider(name = "getStringClaim")
+    public Object[][] getStringClaim() {
+
+        return new Object[][]{
+                {"custom_claim_string_1", "testCustomClaim1", 0},
+                {"custom_claim_string_2", "testCustomClaim2", 1},
+                {"custom_claim_string_3", "testCustomClaim3", 2}
+        };
+    }
+
+    @Test(groups = "wso2.is", description = "Verify the presence of the specified custom string claim in the access " +
+            "token", dataProvider = "getStringClaim")
+    public void testTokenStringClaimAddOperation(String claim, String newClaimValue, int index) throws Exception {
+
+        String claimStr = jwtClaims.getStringClaim(claim);
+        if (index == 0) {
+            Assert.assertEquals(claimStr, newClaimValue);
+        }
```

Comment:
Why only 0th index is checked ?
This will avoid checking custom_claim_string_2 and custom_claim_string_3 right.

Also this is a wrong use of the data provider IMO. data providers are used to pass input data to execute a test case commonly.

Given that the api response is defined why not verify in the code block itself in a static way whether modifications done to claims is available

Reply:
Updated with 4412a6451aab9f4e6c947538138d3c80ad3f36b7